crates/stages/src/trie/mod.rs

```diff
@@ -0,0 +1,467 @@
+use crate::Transaction;
+use bytes::BytesMut;
+use memory_db::{HashKey, MemoryDB};
+use reth_db::{
+    cursor::{DbCursorRO, DbDupCursorRO},
+    database::Database,
+    tables,
+    transaction::DbTx,
+};
+use reth_primitives::{
+    proofs::{KeccakHasher, EMPTY_ROOT},
+    Account, StorageEntry, H256, KECCAK_EMPTY, U256,
+};
+use reth_rlp::{encode_iter, Encodable, RlpDecodable, RlpEncodable};
+use std::{borrow::Borrow, marker::PhantomData};
+use trie_db::{
+    node::{NodePlan, Value},
+    ChildReference, Hasher, NodeCodec, TrieDBMut, TrieDBMutBuilder, TrieLayout, TrieMut,
+};
+
+#[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
+pub(crate) enum TrieError {
+    // TODO: decompose into various different errors
+    #[error("Some error occurred")]
+    InternalError,
+    #[error("{0:?}")]
+    DatabaseError(#[from] reth_db::Error),
+}
+
+struct DBTrieLayout;
+
+impl TrieLayout for DBTrieLayout {
+    const USE_EXTENSION: bool = true;
+
+    const ALLOW_EMPTY: bool = false;
+    const MAX_INLINE_VALUE: Option<u32> = None;
+
+    type Hash = KeccakHasher;
+    type Codec = RLPNodeCodec<Self::Hash>;
+}
+
+// pub struct HashDatabase<DB: Database> {
+//     db: Arc<DB>,
+// }
+
+// impl<H: Hasher, DB: Database, T> HashDB<H, T> for HashDatabase<DB> {
+//     fn get(&self, key: &H::Out, prefix: Prefix<'_>) -> Option<T> {
+//         todo!()
+//     }
+
+//     fn contains(&self, key: &H::Out, prefix: Prefix<'_>) -> bool {
+//         todo!()
+//     }
+
+//     fn insert(&mut self, prefix: Prefix<'_>, value: &[u8]) -> H::Out {
+//         todo!()
+//     }
+
+//     fn emplace(&mut self, key: H::Out, prefix: Prefix<'_>, value: T) {
+//         todo!()
+//     }
+
+//     fn remove(&mut self, key: &H::Out, prefix: Prefix<'_>) {
+//         todo!()
+//     }
+// }
+
+// impl<H: Hasher, T, DB: Database> AsHashDB<H, T> for HashDatabase<DB> {
+//     fn as_hash_db(&self) -> &dyn HashDB<H, T> {
+//         self
+//     }
+
+//     fn as_hash_db_mut<'a>(&'a mut self) -> &'a mut (dyn HashDB<H, T> + 'a) {
+//         self
+//     }
+// }
+
+fn encode_partial(
+    mut partial: impl Iterator<Item = u8>,
+    nibbles: usize,
+    terminating: bool,
+) -> Vec<u8> {
+    debug_assert_ne!(nibbles, 0);
+    let mut out = Vec::with_capacity(nibbles / 2 + 1);
+
+    let mut flag_byte = if terminating { 0x20 } else { 0x00 };
+
+    if nibbles % 2 != 0 {
+        flag_byte |= 0x10;
+        // should never be None
+        flag_byte |= partial.next().unwrap_or_default();
+    }
+    out.push(flag_byte);
+    out.extend(partial);
+    out
+}
+
+#[derive(Debug, Default, Clone)]
+struct RLPNodeCodec<H: Hasher>(PhantomData<H>);
```

Comment:
Would be good to have a comment here, what is an expectation for this, is it only for encoding?

Reply:
Added!